src/engine/Engine.h

```diff
@@ -667,6 +670,64 @@ class Engine {
     }
   }
 
+
+  template<typename A>
+  static void computePatternTrick(const vector<A>* input, vector<array<Id, 2>>* result,
+                                  const vector<PatternID>& hasPattern,
+                                  const CompactStringVector<Id, Id>& hasRelation,
+                                  const CompactStringVector<size_t, Id>& patterns,
+                                  size_t subjectColumn) {
+    ad_utility::HashMap<Id, size_t> predicateCounts;
+    ad_utility::HashMap<size_t, size_t> patternCounts;
+    size_t posInput = 0;
+    size_t subject;
+    size_t lastSubject = ID_NO_VALUE;
+    while (posInput < input->size()) {
+      while ((*input)[posInput][subjectColumn] == lastSubject
+             && posInput < input->size()) {
+        posInput++;
+      }
+      subject = (*input)[posInput][subjectColumn];
+      lastSubject = subject;
+      if (subject < hasPattern.size() && hasPattern[subject] != NO_PATTERN) {
+        // The subject matches a pattern
+        patternCounts[hasPattern[subject]]++;
+      } else if (subject < hasRelation.size()){
+        // The subject does not match a pattern
+
+        std::pair<Id*, size_t> relations = hasRelation[subject];
+        if (relations.second > 0) {
+          for (size_t i = 0; i < relations.second; i++) {
+            auto it = predicateCounts.find(relations.first[i]);
+            if (it == predicateCounts.end()) {
+              predicateCounts[relations.first[i]] =  1;
+            } else {
+              it->second++;
+            }
+          }
+        } else {
+           LOG(TRACE) << "No pattern or has-relation entry found for entity "
+                      << std::to_string(subject) << std::endl;
+        }
+      } else {
+        LOG(TRACE) << "Subject " << subject << " does not appear to be an entity "
+                      "(its id is to high)." << std::endl;
+      }
+      posInput++;
+    }
+    for (const auto& it : patternCounts) {
+      std::pair<Id*, size_t> pattern = patterns[it.first];
+      for (size_t i = 0; i < pattern.second; i++) {
+        predicateCounts[pattern.first[i]] += it.second;
+      }
+    }
+    result->reserve(predicateCounts.size());
+    for (const auto& it : predicateCounts) {
+      result->push_back(array<Id, 2>{it.first, it.second});
```

Comment:
I'm confused here, how is `it.second` of type `Id`? I thought `predicateCounts` maps from `Id` to `size_t`?

Reply:
predicateCounts does map from Id to size_t, the second argument is implicitly converted from size_t to Id (which are currently the same type), to represent the count within the result table (the same way the textual scores are currently stored in the ResultTables when using text operations). I added an explicit static_cast to make it more clear what is being done here.